diff -r 6efca8d4d837 src/network/network_client.cpp
--- src/network/network_client.cpp	Thu May 07 17:08:41 2009 +0000
+++ src/network/network_client.cpp	Fri May 08 20:04:18 2009 +0200
@@ -515,9 +515,15 @@
 			_password_game_seed = p->Recv_uint32();
 			p->Recv_string(_password_server_unique_id, sizeof(_password_server_unique_id));
 			if (MY_CLIENT->HasClientQuit()) return NETWORK_RECV_STATUS_MALFORMED_PACKET;
-
-		case NETWORK_GAME_PASSWORD:
-			ShowNetworkNeedPassword(type);
+	
+			if (_company_pw != NULL) _login_pw = _company_pw;
+			case NETWORK_GAME_PASSWORD:
+				if (_login_pw == NULL) {
+					ShowNetworkNeedPassword(type);
+				} else {
+					SEND_COMMAND(PACKET_CLIENT_PASSWORD)(type, _login_pw);
+					_login_pw = NULL;
+				}
 			return NETWORK_RECV_STATUS_OKAY;
 
 		default: return NETWORK_RECV_STATUS_MALFORMED_PACKET;
diff -r 6efca8d4d837 src/network/network_internal.h
--- src/network/network_internal.h	Thu May 07 17:08:41 2009 +0000
+++ src/network/network_internal.h	Fri May 08 20:04:18 2009 +0200
@@ -123,6 +123,9 @@
 
 void NetworkTCPQueryServer(NetworkAddress address);
 
+extern char *_login_pw;                      // login password from -p argument
+extern char *_company_pw;                    // company password from -P argument
+
 void GetBindAddresses(NetworkAddressList *addresses, uint16 port);
 void NetworkAddServer(const char *b);
 void NetworkRebuildHostList();
diff -r 6efca8d4d837 src/openttd.cpp
--- src/openttd.cpp	Thu May 07 17:08:41 2009 +0000
+++ src/openttd.cpp	Fri May 08 20:04:18 2009 +0200
@@ -158,6 +158,8 @@
 		"  -G seed             = Set random seed\n"
 #if defined(ENABLE_NETWORK)
 		"  -n [ip:port#company]= Start networkgame\n"
+		"  -p password         = Join password\n"
+		"  -P password         = Company password\n"
 		"  -D [ip][:port]      = Start dedicated server\n"
 		"  -l ip[:port]        = Redirect DEBUG()\n"
 #if !defined(__MORPHOS__) && !defined(__AMIGA__) && !defined(WIN32)
@@ -380,6 +382,10 @@
 }
 
 byte _savegame_sort_order;
+#if defined(ENABLE_NETWORK)
+char *_login_pw = NULL;
+char *_company_pw = NULL;
+#endif /* ENABLE_NETWORK */
 #if defined(UNIX) && !defined(__MORPHOS__)
 extern void DedicatedFork();
 #endif
@@ -416,7 +422,7 @@
 	 *   a letter means: it accepts that param (e.g.: -h)
 	 *   a ':' behind it means: it need a param (e.g.: -m<driver>)
 	 *   a '::' behind it means: it can optional have a param (e.g.: -d<debug>) */
-	optformat = "m:s:v:b:hD::n::ei::I:t:d::r:g::G:c:xl:"
+	optformat = "m:s:v:b:hD::n::ei::I:t:d::r:g::G:c:xl:p:P:"
 #if !defined(__MORPHOS__) && !defined(__AMIGA__) && !defined(WIN32)
 		"f"
 #endif
@@ -461,6 +467,12 @@
 		case 'l':
 			debuglog_conn = mgo.opt;
 			break;
+		case 'p':
+			_login_pw = mgo.opt;
+			break;
+		case 'P':
+			_company_pw = mgo.opt;
+			break;
 #endif /* ENABLE_NETWORK */
 		case 'r': ParseResolution(&resolution, mgo.opt); break;
 		case 't': startyear = atoi(mgo.opt); break;
